File: sdk/go/pkg/graphql/graphql.go
/*
 * SPDX-FileCopyrightText: © Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

package graphql

import (
	"errors"
	"fmt"

	"github.com/hypermodeinc/modus/sdk/go/pkg/console"
	"github.com/hypermodeinc/modus/sdk/go/pkg/utils"
)

func Execute[T any](connection, statement string, variables map[string]any) (*Response[T], error) {
	bytes, err := utils.JsonSerialize(variables)
	if err != nil {
		console.Error(err.Error())
		return nil, err
	}

	varsStr := string(bytes)

	response := hostExecuteQuery(&connection, &statement, &varsStr)

	if response == nil {
		return nil, errors.New("failed to execute the GQL query")
	}

	var result Response[T]
	err = utils.JsonDeserialize([]byte(*response), &result)
	if err != nil {
		console.Error(err.Error())
		return nil, err
	}

	if len(result.Errors) > 0 {
		errBytes, err := utils.JsonSerialize(result.Errors)
		if err != nil {
			console.Error(err.Error())
			return nil, err
		}
		console.Error(fmt.Sprint("GraphQL API Errors:" + string(errBytes)))
	}
	return &result, nil
}

type Response[T any] struct {
	Errors []ErrorResult `json:"errors"`
	Data   *T            `json:"data"`
}

type ErrorResult struct {
	Message   string         `json:"message"`
	Locations []CodeLocation `json:"locations"`
	Path      []string       `json:"path"`
}

type CodeLocation struct {
	Line   uint32 `json:"line"`
	Column uint32 `json:"column"`
}

File: sdk/assemblyscript/src/assembly/graphql.ts
/*
 * SPDX-FileCopyrightText: © Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

import { JSON } from "json-as";
import * as utils from "./utils";
import { NamedParams as Variables } from "./database";
export { Variables };

// @ts-expect-error: decorator
@external("modus_graphql_client", "executeQuery")
declare function hostExecuteQuery(
  connection: string,
  statement: string,
  variables: string,
): string;

export function execute<TData>(
  connection: string,
  statement: string,
  variables: Variables = new Variables(),
): Response<TData> {
  const varsJson = variables.toJSON();
  const response = hostExecuteQuery(connection, statement, varsJson);
  if (utils.resultIsInvalid(response)) {
    throw new Error("Error invoking GraphQL API.");
  }

  const results = JSON.parse<Response<TData>>(response);
  if (results.errors) {
    console.error("GraphQL API Errors:" + JSON.stringify(results.errors));
  }
  return results;
}


@json
export class Response<T> {
  errors: ErrorResult[] | null = null;
  data: T | null = null;
  // extensions: Map<string, ???> | null = null;
}


@json
class ErrorResult {
  message!: string;
  locations: CodeLocation[] | null = null;
  path: string[] | null = null;
  // extensions: Map<string, ???> | null = null;
}


@json
class CodeLocation {
  line!: u32;
  column!: u32;
}
